core/src/impl/Kokkos_Core.cpp

```diff
@@ -780,6 +840,7 @@ void Kokkos::Impl::parse_command_line_arguments(
           !kokkos_num_devices_found) {
         num_devices = std::stoi(num1_only);
         settings.set_num_devices(num_devices);
+        settings.set_map_device_id_by("mpi_rank");
```

Comment:
@xZZx @crtrott 
I suspect the issue in #5252 comes from this line.  Probably needs to move it to L815